src/queue.c

```diff
@@ -405,6 +405,62 @@ static int __io_uring_submit(struct io_uring *ring, unsigned submitted,
 	unsigned flags;
 	int ret;
 
+#if defined(__SANITIZE_ADDRESS__)
+	struct io_uring_sq *sq = &ring->sq;
+	unsigned int head;
+	int shift = 0;
+
+	if (ring->flags & IORING_SETUP_SQE128)
+		shift = 1;
+	if (!(ring->flags & IORING_SETUP_SQPOLL))
+		head = *sq->khead;
+	else
+		head = io_uring_smp_load_acquire(sq->khead);
+
+	while(head != sq->sqe_tail)
+	{
+		struct io_uring_sqe *sqe;
+
+		sqe = &sq->sqes[(head & sq->ring_mask) << shift];
+
+		if(sqe->opcode != IORING_OP_URING_CMD && sqe->opcode != IORING_OP_SPLICE && sqe->opcode != IORING_OP_TEE) {
```

Comment:
This whole section is a bit of a mess and will become unwieldy. I think there's two sane ways to do this:

1) Have an array where each opcode can define a sanitize handler to be called before submission.
2) Have an array of structs where each opcode can define which fields to check, and their sizes

That will make it generic and avoid this already incomplete code from becoming very unwieldy down the line as more opcodes are added to it.